# Analysis of Hacker News Posts

Hacker News is a site where user-submitted stories are voted and commented on, similar to Reddit. Posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors. In this project we will examine [data](https://www.kaggle.com/hacker-news/hacker-news-posts) that has been reduced from nearly 300,000 rows to approximately 20,000 rows by removing submissions that did not receive any comments, and randomly sampling from the remaining submissions. Here is a description of the columns:

- id: The unique identifier from Hacker News for the post
- title: title of the post (self explanatory)
- url: the url of the item being linked to
- num_points: the number of upvotes the post received
- num_comments: the number of comments the post received
- author: the name of the account that made the post
- created_at: the date and time the post was made (the time zone is Eastern Time in the US)

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. In this project, we will compare these two types of posts to determine the following:
- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

Let's start by opening the data set and then continue with exploring the data. We remove the header row from the data and assign the header row to the variable *headers*.

In [1]:
### Assign the result to the variable hn
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]

### Display headers and the first five rows
print(headers)
hn[:5]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

## Extracting Ask HN and Show HN posts
Since we're only interested in Ask HN and Show HN posts, we must filter our data. We will create new lists of lists containing just the data for these titles. We will use **str.startswith()** and **str.lower()** to separate these posts.

In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print('Number of posts: ', len(hn))
print('Number of posts in ask_posts: ', len(ask_posts))
print('Number of posts in show_posts: ', len(show_posts))
print('Number of posts in other_posts: ', len(other_posts))

Number of posts:  20100
Number of posts in ask_posts:  1744
Number of posts in show_posts:  1162
Number of posts in other_posts:  17194


We know that of the 20100 posts, 1744 are Ask HN and 1162 are Show HN. Here are the first 5 rows of of the "ask posts" and "show posts" list of lists:

In [4]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [5]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


## Do ask posts or show posts receive more comments?
We will now determine if ask posts or show posts receive more comments on average.

In [10]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += int(row[4])
avg_ask_comments = total_ask_comments / len(ask_posts)

total_show_comments = 0
for row in show_posts:
    total_show_comments += int(row[4])
avg_show_comments = total_show_comments / len(show_posts)

print('The average number of comments on ask posts is', avg_ask_comments)
print('The average number of comments on show posts is', avg_show_comments)

The average number of comments on ask posts is 14.038417431192661
The average number of comments on show posts is 10.31669535283993


On average, ask posts receive nearly 4 more comments than show posts. Since ask posts receive more votes on average, we will focus our remaining analysis on ask posts. 

## Does the time of posting affect comment activity?
Let's determine whether posts created at a certain time are more likely to attract comments. We do this by:
- Calculating the amount of ask posts created in each hour of the day, along with the number of comments receieved.
- Calculating the average number of comments ask posts receive by hour created.

In [12]:
import datetime as dt

result_list = list()
for row in ask_posts:
    result_list.append([row[6], int(row[4])]) # created_at, num_comments

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    # created_at: '9/25/2016 22:57'
    dt_object = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    dt_hour = dt_object.strftime("%H")

    if dt_hour not in counts_by_hour:
        counts_by_hour[dt_hour] = 1
        comments_by_hour[dt_hour] = row[1]
    else:
        counts_by_hour[dt_hour] += 1
        comments_by_hour[dt_hour] += row[1]


We now calculate the average number of comments per post for posts created during each hour of the day.

In [13]:
avg_by_hour = list()
for key in counts_by_hour:
    avg_by_hour.append([key, comments_by_hour[key] / counts_by_hour[key]])

avg_by_hour

[['08', 10.25],
 ['06', 9.022727272727273],
 ['21', 16.009174311926607],
 ['02', 23.810344827586206],
 ['04', 7.170212765957447],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['20', 21.525],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034],
 ['16', 16.796296296296298],
 ['22', 6.746478873239437],
 ['13', 14.741176470588234],
 ['15', 38.5948275862069],
 ['09', 5.5777777777777775],
 ['19', 10.8],
 ['03', 7.796296296296297],
 ['18', 13.20183486238532],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['00', 8.127272727272727],
 ['01', 11.383333333333333],
 ['05', 10.08695652173913],
 ['17', 11.46]]

Let's clean this up to make it more readable.

In [14]:
swap_avg_by_hour = list()
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print('Top 5 Hours for Ask Posts Comments (Time zone: Eastern Time in the US)')
for row in sorted_swap[:5]:
    # format the hours
    str1 = dt.datetime.strptime(row[1], "%H").strftime("%H:%M")
    output = "{hour}: {avg:.2f} average comments per post".format(hour = str1, avg = row[0])
    print(output)

Top 5 Hours for Ask Posts Comments (Time zone: Eastern Time in the US)
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


It appears submitting an ask post between 3pm and 4pm (ET) will yield a greater chance of receiving more comments.